### PR TITLE
#164521891 Implements reset password

### PR DESCRIPTION
**Pull request by valenstical**
**What does this PR do?**
This PR implements reset password feature

**Description of Task to be completed**

- [x] Write unit and integration tests
- [x] Create api endpoint PATCH /api/v1/users/reset_password to handle reset password requests
- [x] Create and mount middleware to validate password parameters and authentication token
- [x] Update usersController class to with function that updates the user password

#### How should this be manually tested

_Prerequisite: Postgres_

1. Run the following commands on terminal:

```bash
git clone -b feature/164521891/reset-password --change-branch \
https://github.com/andela/apollo-ah-backend.git

Navigate to the apollo-ah-backend directory
Run cp .env.example .env
npm install
```

2. Modify `.env` file with your db credentials

3. Run migrations with `npm run migrate`

4. Spin up the server by running `npm run start:dev` on your terminal

5. Make a POST request to `localhost:3000/api/v1/users/signup` with [Postman](https://www.getpostman.com/downloads/). to create an account

6. Make a POST request to `localhost:3000/api/v1/users/forgot_password` with [Postman](https://www.getpostman.com/downloads/) using the email from the account created to request for a new password. Check your email and copy the token parameter from the link provided.

7. Make a POST request to `localhost:3000/api/v1/users/reset_password?token={TOKEN}` with [Postman](https://www.getpostman.com/downloads/) where TOKEN is the token from step 6 above.

❗️ **Important:** Provide user payload when creating an account, sample below:
```json
{
  "email": "example@email.com",
  "password": "secret"
  "username" "user"
}
```
For forgot password endpoint, use the same email from above to make request using this format,
```json
{
  "email": "example@email.com",
}
```
For reset password endpoint, use token gotten from the previous step,
```json
{
  "password": "new password",
  "confirmPassword": "new password"
}
```

**NOTE:** To run test locally, create `.env.test` file. Run command: `cp .env.example .env.test`

#### Any background context you want to provide
N/A

#### What are the relevant pivotal tracker stories

[#164521891](https://www.pivotaltracker.com/story/show/164521891)

#### Screenshots
N/A

_-- Template format by @devdbrandy_

### CODE REVIEW

File: controllers/usersController.js [hunk 1]
```diff
@@ -73,7 +73,7 @@ class UsersController {
   static async sendPasswordRecoveryLink(request, response, next) {
     try {
       const { email } = request.body;
-      const token = jwt.sign({ email }, env('APP_KEY'), { expiresIn: '1h' });
+      const token = jwt.sign({ email }, env('PASSWORD_RESET_KEY'), { expiresIn: '1h' });
```
Comment (devdbrandy) [inline controllers/usersController.js]: Hi @valenstical You doing great. Just to clarify, why is there additional secret key for encrypting the email payload, whereas an `APP_KEY` is already available?

Comment (valenstical) [reply controllers/usersController.js]: Hello @devdbrandy , 
Due to the sensitivity involved in resetting a password, it is advisable to use a different private key with JWT. This way,  a compromise on the jswon web token from other parts of the application using JWT with the  APP_KEY will not have any negative effect on resetting password.

File: controllers/usersController.js [hunk 2]
```diff
@@ -110,6 +110,37 @@ class UsersController {
...
+ * @memberof UsersController
+ */
+  static async resetPassword(request, response, next) {
+    try {
+      const count = await User.update(
+        {
+          password: request.body.password
+        },
+        {
+          where: { email: request.recoveryEmail }
```
Comment (aanchirinah) [inline controllers/usersController.js]: #### Nitpick
- `request.recoveryEmail`?

File: middlewares/validator.js [hunk 3]
```diff
@@ -17,7 +17,7 @@ export default class Validator {
   static validateRegistration() {
     return [
       ...Validator.validateEmail(),
```
Comment (aanchirinah) [inline middlewares/validator.js]: #### Nitpick
- How will this look like? Can I see an example of this?

File: controllers/usersController.js (hunk 2 above)
Comment (valenstical) [reply controllers/usersController.js]: I have refactored this to include the recovery email in the body object instead.

File: middlewares/validator.js (hunk 3 above)
Comment (valenstical) [reply middlewares/validator.js]: Screenshot 2019-03-14 at 4 19 06 AM

Comment (valenstical) [reply middlewares/validator.js]: This is used with express-validator, which requires a validation error catching middleware and an array of checks.

File: controllers/usersController.js (hunk 1 above)
Comment (devdbrandy) [reply controllers/usersController.js]: Okay, that makes lot of sense. Great job @valenstical.

File: middlewares/validator.js (hunk 3 above)
Comment (aanchirinah) [reply middlewares/validator.js]: @valenstical , I meant how will the return array look like.

**Pull request merged**